src/components/Modal/Modal.tsx

```diff
@@ -11,54 +11,97 @@ import {
 
-export interface BaseModalProps {
+export interface BaseProps {
   /** The content to display in the modal body. */
   bodyContent?: string | JSX.Element;
-  /** The label for the close button. */
-  closeButtonLabel?: string;
   /** The text to display in the modal heading, can be a string or JSX Element. */
   headingText?: string | JSX.Element;
   /** ID that other components can cross reference for accessibility purposes. */
   id?: string;
   /** Boolean to determine if the modal is open or closed. */
   isOpen?: boolean;
+}
+
+export interface ConfirmationModalProps {
+  /** The label for the confirm button. */
+  confirmButtonLabel?: string;
+  /* Function to call when the modal action is canceled. */
+  onCancel: () => void;
+  /* Function to call when the modal action is confirmed. */
+  onConfirm: () => void;
+  /** The `Modal` variant to render. */
+  type: "confirmation";
+  /** The label for the close button. This prop is used for the
+   * "cancel" button in the confirmation variant. */
+  closeButtonLabel?: string;
+  /* Function to call when the modal is closed. This prop is not 
+  used in the confirmation variant. */
+  onClose?: never;
+}
+
+export interface DefaultModalProps {
+  /** The label for the close button. */
+  closeButtonLabel?: string;
   /* Function to call when the modal is closed. */
   onClose?: () => void;
+  /** The `Modal` variant to render. */
+  type: "default";
+  /** The label for the confirm button. This prop is not used
+   * in the default variant. */
+  confirmButtonLabel?: never;
+  /* Function to call when the modal action is canceled. This prop 
+  is not used in the default variant. */
+  onCancel?: never;
+  /* Function to call when the modal action is confirmed. This prop is not 
+    used in the default variant. */
+  onConfirm?: never;
 }
 
+// We want either the default or the confirmation props.
+export type ModalTypeProps = ConfirmationModalProps | DefaultModalProps;
+// And here combine the special types with the base props.
+export type BaseModalProps = BaseProps & ModalTypeProps;
+
 export interface ModalProps {
   /** The text to display on the button that opens the modal. */
   buttonText?: string;
   /** ID that other components can cross reference for accessibility purposes */
   id?: string;
   /** Props to update the internal `Modal` component. This contains the
-   * `bodyContent`, `closeButtonLabel`, `headingText`, `isOpen`, and
-   * `onClose` props. */
+   * `bodyContent`, `headingText`, `isOpen`, and the modal type props. */
   modalProps: BaseModalProps;
 }
 
+// Type guard for the `Modal` variant.
+export function isDefaultType(type: BaseModalProps["type"]): type is "default" {
+  return type === "default";
+}
+
 export const BaseModal: ChakraComponent<
   React.ForwardRefExoticComponent<
-    React.PropsWithChildren<BaseModalProps> &
-      React.RefAttributes<HTMLButtonElement>
+    BaseModalProps & React.RefAttributes<HTMLButtonElement>
```

Comment:
This is not necessarily related to your changes, but this component is using the `forwardRef` function but the `ref` prop is not defined. So in this case, the second part of `BaseModalProps & React.RefAttributes...` is not doing anything. You could add a ref on line 91, `(props: BaseModalProps, ref: ...)`, but I'm not sure if you can pass it to the `ChakraModal` or if that is even the correct element to pass this ref to. Perhaps this is the PR where the `forwardRef` can just be removed.